---
# _Introduction to Pandas_
## _Date: 27 Sep_
---

In [1]:
import pandas as pd

In [2]:
ckd_data = pd.read_csv('ChronicKidneyDisease.csv')
ckd_data.head()

id   age    bp     sg   al   su     rbc        pc         pcc          ba  \
0   0  48.0  80.0  1.020  1.0  0.0     NaN    normal  notpresent  notpresent   
1   1   7.0  50.0  1.020  4.0  0.0     NaN    normal  notpresent  notpresent   
2   2  62.0  80.0  1.010  2.0  3.0  normal    normal  notpresent  notpresent   
3   3  48.0  70.0  1.005  4.0  0.0  normal  abnormal     present  notpresent   
4   4  51.0  80.0  1.010  2.0  0.0  normal    normal  notpresent  notpresent   

   ...  pcv    wc   rc  htn   dm  cad appet   pe  ane classification  
0  ...   44  7800  5.2  yes  yes   no  good   no   no            ckd  
1  ...   38  6000  NaN   no   no   no  good   no   no            ckd  
2  ...   31  7500  NaN   no  yes   no  poor   no  yes            ckd  
3  ...   32  6700  3.9  yes   no   no  poor  yes  yes            ckd  
4  ...   35  7300  4.6   no   no   no  good   no   no            ckd  

[5 rows x 26 columns]

### Q1: 
Which of the following is the correct syntax to display the first 10 records & last 7 records from pandas dataframe?
> **a) ckd_data.head(10) & ckd_data.tail(7)**  
> b) ckd_data.head(9) & ckd_data.tail(6)  
> c) ckd_data.head(11) & ckd_data.tail(8)  
> d) None of the above

In [3]:
ckd_data.head(10) 

id   age     bp     sg   al   su       rbc        pc         pcc  \
0   0  48.0   80.0  1.020  1.0  0.0       NaN    normal  notpresent   
1   1   7.0   50.0  1.020  4.0  0.0       NaN    normal  notpresent   
2   2  62.0   80.0  1.010  2.0  3.0    normal    normal  notpresent   
3   3  48.0   70.0  1.005  4.0  0.0    normal  abnormal     present   
4   4  51.0   80.0  1.010  2.0  0.0    normal    normal  notpresent   
5   5  60.0   90.0  1.015  3.0  0.0       NaN       NaN  notpresent   
6   6  68.0   70.0  1.010  0.0  0.0       NaN    normal  notpresent   
7   7  24.0    NaN  1.015  2.0  4.0    normal  abnormal  notpresent   
8   8  52.0  100.0  1.015  3.0  0.0    normal  abnormal     present   
9   9  53.0   90.0  1.020  2.0  0.0  abnormal  abnormal     present   

           ba  ...  pcv     wc   rc  htn   dm  cad appet   pe  ane  \
0  notpresent  ...   44   7800  5.2  yes  yes   no  good   no   no   
1  notpresent  ...   38   6000  NaN   no   no   no  good   no   no   
2  notpresent  ...   31   7500  NaN   no  yes   no  poor   no  yes   
3  notpresent  ...   32   6700  3.9  yes   no   no  poor  yes  yes   
4  notpresent  ...   35   7300  4.6   no   no   no  good   no   no   
5  notpresent  ...   39   7800  4.4  yes  yes   no  good  yes   no   
6  notpresent  ...   36    NaN  NaN   no   no   no  good   no   no   
7  notpresent  ...   44   6900    5   no  yes   no  good  yes   no   
8  notpresent  ...   33   9600  4.0  yes  yes   no  good   no  yes   
9  notpresent  ...   29  12100  3.7  yes  yes   no  poor   no  yes   

  classification  
0            ckd  
1            ckd  
2            ckd  
3            ckd  
4            ckd  
5            ckd  
6            ckd  
7            ckd  
8            ckd  
9            ckd  

[10 rows x 26 columns]

In [4]:
ckd_data.tail(7)

id   age    bp     sg   al   su     rbc      pc         pcc          ba  \
393  393  43.0  60.0  1.025  0.0  0.0  normal  normal  notpresent  notpresent   
394  394  50.0  80.0  1.020  0.0  0.0  normal  normal  notpresent  notpresent   
395  395  55.0  80.0  1.020  0.0  0.0  normal  normal  notpresent  notpresent   
396  396  42.0  70.0  1.025  0.0  0.0  normal  normal  notpresent  notpresent   
397  397  12.0  80.0  1.020  0.0  0.0  normal  normal  notpresent  notpresent   
398  398  17.0  60.0  1.025  0.0  0.0  normal  normal  notpresent  notpresent   
399  399  58.0  80.0  1.025  0.0  0.0  normal  normal  notpresent  notpresent   

     ...  pcv    wc   rc  htn  dm  cad appet  pe ane classification  
393  ...   54  7400  5.4   no  no   no  good  no  no         notckd  
394  ...   45  9500  4.6   no  no   no  good  no  no         notckd  
395  ...   47  6700  4.9   no  no   no  good  no  no         notckd  
396  ...   54  7800  6.2   no  no   no  good  no  no         notckd  
397  ...   49  6600  5.4   no  no   no  good  no  no         notckd  
398  ...   51  7200  5.9   no  no   no  good  no  no         notckd  
399  ...   53  6800  6.1   no  no   no  good  no  no         notckd  

[7 rows x 26 columns]

### Q2:
Display the data types of the given dataset and answer - which all data types are there in the dataset and how many numbers of each data type are present?
> a) int64 (1), float64 (20), object (14)  
> **b) int64 (1), float64 (11), object (14)**  
> c) float64 (11), object (14)  
> d) int64 (1), float64 (11), object (16)

In [5]:
ckd_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 26 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              400 non-null    int64  
 1   age             391 non-null    float64
 2   bp              388 non-null    float64
 3   sg              353 non-null    float64
 4   al              354 non-null    float64
 5   su              351 non-null    float64
 6   rbc             248 non-null    object 
 7   pc              335 non-null    object 
 8   pcc             396 non-null    object 
 9   ba              396 non-null    object 
 10  bgr             356 non-null    float64
 11  bu              381 non-null    float64
 12  sc              383 non-null    float64
 13  sod             313 non-null    float64
 14  pot             312 non-null    float64
 15  hemo            348 non-null    float64
 16  pcv             330 non-null    object 
 17  wc              295 non-null    obj

### Q3:
See the summary of the dataset and answer -  what is the frequency (freq) of the top value in the classification column and what is that top value? 
> a) 246, notckd  
> **b) 248, ckd**  
> c) NaN  
> d) None of the above

In [6]:
ckd_data['classification'].describe()

count     400
unique      3
top       ckd
freq      248
Name: classification, dtype: object

### Q4:
Select the correct syntax for creating a dataframe with - (Rows: 247 to 253) and (Columns: id, age, classification)
> a) DataFrame_Name.iloc[200:205, ["id", "age", "classification"]]    
> b) DataFrame_Name.loc[247:254, ["id", "age", "classification"]]  
> c) DataFrame_Name.iloc[[247:254],["id", "age", "classification"]]  
> **d) DataFrame_Name.loc[247:253, ["id", "age", "classification"]]**  

In [7]:
ckd_data.loc[247:253, ["id", "age", "classification"]]

id   age classification
247  247  54.0            ckd
248  248  59.0            ckd
249  249  56.0            ckd
250  250  40.0         notckd
251  251  23.0         notckd
252  252  45.0         notckd
253  253  57.0         notckd

### Q5:
What’s the highest age in the dataset?
> **a) 90.0**  
> b) 80.0  
> c) 83.0  
> d) 51.0  

In [8]:
ckd_data['age'].max()

90.0

### Q6:
What will be the output/result when we execute the following code? 
```
DataFrame.rename(columns = {
   "id": "Id",
   "age": "Age",
   "bp": "blood_pressure",
}, inplace=True)
```
> a) The code renames column id, age, bp and creates a new dataframe with only newly named columns.  
> b) The code renames id, age, bp as Id, Age, blood_pressure.  
> c) The code renames column index 0, 1, 2.   
> **d) Both B and C**  
> e) Doesn't renames columns because inplace should be False. 

In [9]:
ckd_data.columns

Index(['id', 'age', 'bp', 'sg', 'al', 'su', 'rbc', 'pc', 'pcc', 'ba', 'bgr',
       'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wc', 'rc', 'htn', 'dm', 'cad',
       'appet', 'pe', 'ane', 'classification'],
      dtype='object')

In [10]:
ckd_data.rename(columns = {
   "id": "Id",
   "age": "Age",
   "bp": "blood_pressure",
}, inplace=True)

In [11]:
ckd_data.columns

Index(['Id', 'Age', 'blood_pressure', 'sg', 'al', 'su', 'rbc', 'pc', 'pcc',
       'ba', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wc', 'rc', 'htn',
       'dm', 'cad', 'appet', 'pe', 'ane', 'classification'],
      dtype='object')

### Q7:
How many values are missing in Id, Age, and classification column respectively?
> a) 9, 12, 0  
> b) 0, 9, 12  
> **c) 0, 9, 0**  
> d) These columns have no missing values

In [12]:
ckd_data[['Id', 'Age', 'classification']].isna().sum()

Id                0
Age               9
classification    0
dtype: int64

### Q8:
Your task is to fill the missing values of the blood_pressure column with its median value. Don’t forget to inplace and summarize the changes in the blood_pressure column using describe() method. After filling the missing values, what do you observe? 
> a) The mean has increased slightly.  
> b) The standard deviation (std) has decreased.  
> **c) Both A & B**  
> d) Datatype has been changed. 

In [13]:
ckd_data['blood_pressure'].isna().sum()

12

In [14]:
ckd_data['blood_pressure'].describe()

count    388.000000
mean      76.469072
std       13.683637
min       50.000000
25%       70.000000
50%       80.000000
75%       80.000000
max      180.000000
Name: blood_pressure, dtype: float64

In [15]:
bp_median = ckd_data['blood_pressure'].median()
ckd_data['blood_pressure'].fillna(bp_median, inplace=True)

In [16]:
ckd_data['blood_pressure'].describe()

count    400.000000
mean      76.575000
std       13.489785
min       50.000000
25%       70.000000
50%       80.000000
75%       80.000000
max      180.000000
Name: blood_pressure, dtype: float64

### Q9:
Try to fill the missing values of the cad column with its mean value. After executing your code, what do you observe?
> a) The missing values were filled by the mean value.  
> b) The missing values were filled by the median value.  
> c) Both A & B  
> **d) Error performing the operation.**

In [17]:
ckd_data['cad'].isna().sum()

2

In [18]:
ckd_data['cad'].fillna(ckd_data['cad'].mean(), inplace=True)

TypeError: can only concatenate str (not "int") to str

### Q10:
Which of the following can be used to export a sample dataframe with the index numbers?
> a) Sample_DataFrame_Name.to_csv("test.csv", index=False)  
> b) Sample_DataFrame_Name.to_csv("test.csv", index=True)  
> c) Sample_DataFrame_Name.to_csv("test.csv")  
> **d) Both B & C**

In [19]:
ckd_data.to_csv("test.csv", index=False)
pd.read_csv("test.csv").head()

Id   Age  blood_pressure     sg   al   su     rbc        pc         pcc  \
0   0  48.0            80.0  1.020  1.0  0.0     NaN    normal  notpresent   
1   1   7.0            50.0  1.020  4.0  0.0     NaN    normal  notpresent   
2   2  62.0            80.0  1.010  2.0  3.0  normal    normal  notpresent   
3   3  48.0            70.0  1.005  4.0  0.0  normal  abnormal     present   
4   4  51.0            80.0  1.010  2.0  0.0  normal    normal  notpresent   

           ba  ...  pcv    wc   rc  htn   dm  cad appet   pe  ane  \
0  notpresent  ...   44  7800  5.2  yes  yes   no  good   no   no   
1  notpresent  ...   38  6000  NaN   no   no   no  good   no   no   
2  notpresent  ...   31  7500  NaN   no  yes   no  poor   no  yes   
3  notpresent  ...   32  6700  3.9  yes   no   no  poor  yes  yes   
4  notpresent  ...   35  7300  4.6   no   no   no  good   no   no   

  classification  
0            ckd  
1            ckd  
2            ckd  
3            ckd  
4            ckd  

[5 rows x 26 columns]

In [20]:
ckd_data.to_csv("test.csv", index=True)
pd.read_csv("test.csv").head()

Unnamed: 0  Id   Age  blood_pressure     sg   al   su     rbc        pc  \
0           0   0  48.0            80.0  1.020  1.0  0.0     NaN    normal   
1           1   1   7.0            50.0  1.020  4.0  0.0     NaN    normal   
2           2   2  62.0            80.0  1.010  2.0  3.0  normal    normal   
3           3   3  48.0            70.0  1.005  4.0  0.0  normal  abnormal   
4           4   4  51.0            80.0  1.010  2.0  0.0  normal    normal   

          pcc  ... pcv    wc   rc  htn   dm  cad  appet   pe  ane  \
0  notpresent  ...  44  7800  5.2  yes  yes   no   good   no   no   
1  notpresent  ...  38  6000  NaN   no   no   no   good   no   no   
2  notpresent  ...  31  7500  NaN   no  yes   no   poor   no  yes   
3     present  ...  32  6700  3.9  yes   no   no   poor  yes  yes   
4  notpresent  ...  35  7300  4.6   no   no   no   good   no   no   

  classification  
0            ckd  
1            ckd  
2            ckd  
3            ckd  
4            ckd  

[5 rows x 27 columns]

In [21]:
ckd_data.to_csv("test.csv")
pd.read_csv("test.csv").head()

Unnamed: 0  Id   Age  blood_pressure     sg   al   su     rbc        pc  \
0           0   0  48.0            80.0  1.020  1.0  0.0     NaN    normal   
1           1   1   7.0            50.0  1.020  4.0  0.0     NaN    normal   
2           2   2  62.0            80.0  1.010  2.0  3.0  normal    normal   
3           3   3  48.0            70.0  1.005  4.0  0.0  normal  abnormal   
4           4   4  51.0            80.0  1.010  2.0  0.0  normal    normal   

          pcc  ... pcv    wc   rc  htn   dm  cad  appet   pe  ane  \
0  notpresent  ...  44  7800  5.2  yes  yes   no   good   no   no   
1  notpresent  ...  38  6000  NaN   no   no   no   good   no   no   
2  notpresent  ...  31  7500  NaN   no  yes   no   poor   no  yes   
3     present  ...  32  6700  3.9  yes   no   no   poor  yes  yes   
4  notpresent  ...  35  7300  4.6   no   no   no   good   no   no   

  classification  
0            ckd  
1            ckd  
2            ckd  
3            ckd  
4            ckd  

[5 rows x 27 columns]

---